### PR TITLE
Enforce CPU time limit in SqlQueryManager

### PR DESCRIPTION
**Pull request by yangyangyyy**
logic is very similar to existing ClusterMemoryManager. adding necessary config hooks

### CODE REVIEW

**Comment by cberner** [comment]
Looks like you're missing some tests. See Travis

**Comment by electrum** [comment]
Please don't use passive voice for commit messages.

File: presto-main/src/main/java/com/facebook/presto/ExceededCpuLimitException.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import com.facebook.presto.spi.PrestoException;
+
+import static com.facebook.presto.spi.StandardErrorCode.EXCEEDED_MEMORY_LIMIT;
+
+public class ExceededCpuLimitException
+        extends PrestoException
+{
+    public ExceededCpuLimitException(long millis)
+    {
+        super(EXCEEDED_MEMORY_LIMIT, String.format("Exceeded cpu limit of :%d milliseconds", millis));
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/ExceededCpuLimitException.java]: Static import

File: presto-main/src/main/java/com/facebook/presto/ExceededCpuLimitException.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+package com.facebook.presto;
+
+import com.facebook.presto.spi.PrestoException;
+
+import static com.facebook.presto.spi.StandardErrorCode.EXCEEDED_MEMORY_LIMIT;
+
+public class ExceededCpuLimitException
+        extends PrestoException
+{
+    public ExceededCpuLimitException(long millis)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/ExceededCpuLimitException.java]: I think this should take an Airlift `Duration`

File: presto-main/src/main/java/com/facebook/presto/ExceededCpuLimitException.java (hunk 1 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/ExceededCpuLimitException.java]: Remove the `:`

Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/ExceededCpuLimitException.java]: Capitalize "CPU"

File: presto-tests/src/test/java/com/facebook/presto/memory/TestMemoryManager.java [hunk 3]
```diff
@@ -243,6 +243,18 @@ public void testQueryMemoryLimit()
         }
     }
 
+    @Test(timeOut = 240_000, expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "Exceeded cpu limit of :.*")
```
Comment (haozhun) [inline presto-tests/src/test/java/com/facebook/presto/memory/TestMemoryManager.java]: Test fails. The message here does not match the message actually thrown (colon). See Travis.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 4]
```diff
@@ -272,6 +272,12 @@ public long getTotalMemoryReservation()
                 .sum();
     }
 
+    public long getTotalCpuTime()
+    {
+        return stages.values().stream()
+                .mapToLong(SqlStageExecution::getCpuTime)
+                .sum();
+    }
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: blank line after

**Comment by haozhun** [comment]
@yangyangyyy Please address my comments, and also rebase on latest trunk. Let me know when you finish.

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryManagerConfig.java [hunk 5]
```diff
@@ -34,6 +35,7 @@
     private DataSize maxQueryMemoryPerNode = new DataSize(1, GIGABYTE);
     private boolean killOnOutOfMemory;
     private Duration killOnOutOfMemoryDelay = new Duration(5, MINUTES);
+    private Duration maxQueryCpuTime = new Duration(100, DAYS);
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryManagerConfig.java]: Make this `Long.MAX_VALUE, NANOSECONDS`. 100 days is half an hour in a large cluster. It's too small.

File: presto-tests/src/test/java/com/facebook/presto/memory/TestMemoryManager.java [hunk 6]
```diff
@@ -280,6 +280,18 @@ public void testQueryMemoryLimit()
...
+            throws Exception
+    {
+        Map<String, String> properties = ImmutableMap.<String, String>builder()
+                .put("query.max-cpu-time", "1ms")
+                .build();
+        try (QueryRunner queryRunner = createQueryRunner(SESSION, properties)) {
+            // the following test query is known to run for a long time
+            queryRunner.execute(SESSION, "SELECT COUNT(*), clerk FROM orders GROUP BY clerk");
+        }
+    }
```
Comment (haozhun) [inline presto-tests/src/test/java/com/facebook/presto/memory/TestMemoryManager.java]: blank line after

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryManagerConfig.java (hunk 5 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryManagerConfig.java]: Remember to change the corresponding test.

**Comment by martint** [comment]
@cberner, can you do a final pass on this one?

**Comment by cberner** [comment]
Merged. Thanks!

**Pull request closed without merging**